File: platforms/ios/Sources/WhiskerRuntime/module/HostModuleDispatcher.swift
import WhiskerModule

/// Routes runtime module calls through the Host-local module registry.
final class HostModuleDispatcher {
    func invoke(
        module name: String,
        method: String,
        rawArgs: UnsafePointer<WhiskerValueRaw>?,
        argumentCount: Int,
        isAsync: Bool,
        result: @escaping WhiskerModuleResult,
        resultData: UnsafeMutableRawPointer?
    ) -> Bool {
        guard let module = WhiskerModuleRegistry.module(named: name) else {
            deliver(.error("module not registered: \(name)"), result, resultData)
            return true
        }
        let args = WhiskerValue.decodeArray(rawArgs, count: argumentCount)
        let settle: (WhiskerValue) -> Void = { [weak self] value in
            self?.deliver(value, result, resultData)
        }
        if isAsync {
            let promise = WhiskerPromise(onSettle: settle)
            if module.dispatchModuleFunctionAsync(method, args, promise) { return true }
        }
        settle(module.dispatchModuleFunction(method, args))
        return true
    }

    func observe(owner: AnyObject, module: String, event: String, observing: Bool) {
        WhiskerModuleEventCenter.setObserving(
            owner: owner,
            module: module,
            event: event,
            observing: observing
        )
    }

    private func deliver(
        _ value: WhiskerValue,
        _ result: WhiskerModuleResult,
        _ resultData: UnsafeMutableRawPointer?
    ) {
        var raw = value.toRaw()
        result(resultData, &raw)
        WhiskerValue.releaseRaw(&raw)
    }
}

File: platforms/android/runtime/src/main/kotlin/rs/whisker/runtime/module/HostModuleDispatcher.kt
package rs.whisker.runtime.module

import rs.whisker.runtime.WhiskerModuleEventCenter
import rs.whisker.runtime.WhiskerModuleRegistry
import rs.whisker.runtime.WhiskerValue

/** Bridges typed module invocations to the Host-local module registry. */
internal class HostModuleDispatcher(
    private val resolve: (Long, Long, WhiskerValue) -> Unit,
) {
    fun invoke(
        module: String,
        method: String,
        args: Array<WhiskerValue>,
        isAsync: Boolean,
        callbackPtr: Long,
        userDataPtr: Long,
    ): Boolean = try {
        val settle: (WhiskerValue) -> Unit = { value ->
            resolve(callbackPtr, userDataPtr, value)
        }
        if (isAsync && WhiskerModuleRegistry.invokeDispatchAsync(module, method, args, settle)) {
            true
        } else {
            settle(WhiskerModuleRegistry.invokeDispatch(module, method, args))
            true
        }
    } catch (error: Throwable) {
        resolve(
            callbackPtr,
            userDataPtr,
            WhiskerValue.Err(
                "module $module.$method failed: ${error.message ?: error.javaClass.simpleName}",
            ),
        )
        true
    }

    fun observe(owner: Any, module: String, event: String, observing: Boolean) {
        WhiskerModuleEventCenter.setObserving(owner, module, event, observing)
    }
}
